Fix when starting apptainer failed with `instance --fakeroot`

JasonYangShadow opened the pull request:
## Description of the Pull Request (PR):

current situation is that we have a trial fix but it is not perfect because of other issues.
1. fakeroot mode is problematic for alpine distros, should work fine for other distros
```
[vagrant@localhost test]$ apptainer instance start --fakeroot --ignore-subuid docker://ubuntu:latest a1
INFO:    Using cached SIF image
INFO:    User not listed in /etc/subuid, trying root-mapped namespace
INFO:    Using cached SIF image
INFO:    Using fakeroot command combined with root-mapped namespace
INFO:    Instance stats will not be available - requires cgroups v2 with systemd as manager.
INFO:    squashfuse not found, will not be able to mount SIF or other squashfs files
INFO:    gocryptfs not found, will not be able to use gocryptfs
INFO:    Converting SIF file to temporary sandbox...
INFO:    instance started successfully
```

2. issue arises in this fix, because the created instance is not managed corrected by the CLI, as inside the newly created namespace, the user becomes root, and all instance info is saved into the root path rather than the normal user path ( ~/.apptainer/instances/app/localhost.localdomain/root rather than  ~/.apptainer/instances/app/localhost.localdomain/$USER). As shown above the instance is started correctly, but can not be listed by `apptainer instance list` command.  
```
[vagrant@localhost test]$ apptainer instance list
INSTANCE NAME    PID    IP    IMAGE
```
3. So I have to disable the root-mapped, which will make the uid to 0 inside container. 


### This fixes or addresses the following GitHub issues:

 - Fixes #2189


#### Before submitting a PR, make sure you have done the following:

- Read the [Guidelines for Contributing](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/apptainer/apptainer/blob/main/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added tests to validate this PR, linted with `make check`  and tested this PR locally with a `make test`, and `make testall` if possible (see CONTRIBUTING.md).
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/apptainer/apptainer/blob/main/CONTRIBUTORS.md)

DrDaveD commented on CHANGELOG.md:
```diff
@@ -25,6 +25,7 @@ For older changes see the [archived Singularity change log](https://github.com/a
 - Fixed a segmentation violation issue when running Apptainer checkpoint.
 - Added apparmor profiles for ubuntu 24.04 or higher distros.
 - Fixed an issue that Apptainer won't read default docker credentials.
+- Trial fix when starting apptainer failed with `instance --fakeroot`.
```
The message needs to be in the Changes for v1.3.x section, not in the v1.3.3 section.

I also don't understand what you mean by "Trial fix".  You don't know whether or not it is going to work?  If not I would still advise making the message be that it is a fix, and before merging we'll ask the user to to test it.  Also the message should better describe what kind of failure it was.

DrDaveD commented on internal/pkg/runtime/launch/launcher_linux.go:
```diff
@@ -103,7 +103,7 @@ func (l *Launcher) Exec(ctx context.Context, image string, args []string, instan
 			// Already running root-mapped unprivileged
 			l.cfg.Fakeroot = false
-			l.cfg.Namespaces.User = true
+			l.cfg.Namespaces.User = false
```
This is confusing.  If it is unprivileged, why would user namespaces be false?

JasonYangShadow replied:
> This is confusing. If it is unprivileged, why would user namespaces be false?

this is a kind of tricky fix, there are some procedures that still unknown to me, as listed here
https://github.com/apptainer/apptainer/pull/2329#issuecomment-2224195177

DrDaveD replied:
The answer is that setting this to false prevents Apptainer from allocating an additional user namespace, which is not only unnecessary, it interferes with an instance running with `--fakeroot`.  I think it would be good to add a comment about that here.

JasonYangShadow replied:
yes, I added detailed comments.

JasonYangShadow commented:
without this line
https://github.com/apptainer/apptainer/blob/76d999f59a88c146ee04ee5e59bf72647c577f05/internal/pkg/runtime/launch/launcher_linux.go#L1090
with root-mapped mode, apptainer will create instance file under incorrect path, (root instead of $USER).

We can easily test with the either of the following commands
```
unshare -r apptainer instance start docker://ubuntu:latest a1
unshare -r apptainer instance start --fakeroot docker://ubuntu:latest a1
apptainer instance start --fakeroot --ignore-subuid docker://ubuntu:latest a1
```
we can always see that the instance is created successfully, but it is created under ` ~/.apptainer/instances/app/localhost.localdomain/root`, this is because `namespace.HostUID` will return `root` in this mode.

JasonYangShadow commented:
@DrDaveD 
I pushed some code changes, this somehow fails this e2e test
https://github.com/JasonYangShadow/apptainer/blob/56d2828b6ae8f04fa70ace514ab078db482b6239/e2e/run/run.go#L388
```
2024-07-22T13:34:57.6112152Z === NAME  TestE2E/PAR/RUN/add_package_with_fakeroot_and_tmpfs
2024-07-22T13:34:57.6113041Z     apptainercmd.go:694: 
2024-07-22T13:34:57.6116253Z         "/usr/local/bin/apptainer exec --fakeroot --userns --overlay /tmp/stest.1614578294/3998688929/overlaymode31293147936 --ignore-subuid /tmp/stest.1614578294/test-debian.sif sh -c apt-get update && apt-get install -y openssh-client" output:
2024-07-22T13:34:57.6119425Z         INFO:    User not listed in /etc/subuid, trying root-mapped namespace
2024-07-22T13:34:57.6120808Z         E: setgroups 65534 failed - setgroups (1: Operation not permitted)
2024-07-22T13:34:57.6122064Z         E: setegid 65534 failed - setegid (22: Invalid argument)
2024-07-22T13:34:57.6124042Z         E: seteuid 100 failed - seteuid (22: Invalid argument)
2024-07-22T13:34:57.6125269Z         E: setgroups 0 failed - setgroups (1: Operation not permitted)
2024-07-22T13:34:57.6127237Z         W: chown to _apt:root of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (22: Invalid argument)
2024-07-22T13:34:57.6129577Z         W: chown to _apt:root of directory /var/lib/apt/lists/auxfiles failed - SetupAPTPartialDirectory (22: Invalid argument)
2024-07-22T13:34:57.6131335Z         E: setgroups 65534 failed - setgroups (1: Operation not permitted)
2024-07-22T13:34:57.6132573Z         E: setegid 65534 failed - setegid (22: Invalid argument)
2024-07-22T13:34:57.6133752Z         E: seteuid 100 failed - seteuid (22: Invalid argument)
2024-07-22T13:34:57.6134928Z         E: setgroups 0 failed - setgroups (1: Operation not permitted)
2024-07-22T13:34:57.6136716Z         E: Method gave invalid 400 URI Failure message: Failed to setgroups - setgroups (1: Operation not permitted)
2024-07-22T13:34:57.6138018Z         E: Method http has died unexpectedly!
2024-07-22T13:34:57.6138995Z         E: Sub-process http returned an error code (112)
2024-07-22T13:34:57.6139777Z         Reading package lists...
2024-07-22T13:34:57.6140300Z         
2024-07-22T13:34:57.6141292Z     apptainercmd.go:694: got 100 as exit code and was expecting 0: exit status 100
2024-07-22T13:34:57.6144940Z         waiting for command "/usr/local/bin/apptainer exec --fakeroot --userns --overlay /tmp/stest.1614578294/3998688929/overlaymode31293147936 --ignore-subuid /tmp/stest.1614578294/test-debian.sif sh -c apt-get update && apt-get install -y openssh-client"
2024-07-22T13:34:57.6147901Z         github.com/apptainer/apptainer/e2e/internal/e2e.TestEnv.RunApptainer.func1
2024-07-22T13:34:57.6149499Z         	github.com/apptainer/apptainer/e2e/internal/e2e/apptainercmd.go:669
2024-07-22T13:34:57.6151263Z         github.com/apptainer/apptainer/e2e/internal/e2e.TestEnv.RunApptainer.func2
2024-07-22T13:34:57.6152947Z         	github.com/apptainer/apptainer/e2e/internal/e2e/apptainercmd.go:694
2024-07-22T13:34:57.6153850Z         runtime.goexit
2024-07-22T13:34:57.6154521Z         	runtime/asm_amd64.s:1650
```

not sure if it is expected, but will take a look at it tomorrow. If you have any comments, feel free to share here.

DrDaveD commented:
I think I know what happened.  The removal of the `--fakeroot` flag resulted in it only doing the `unshare -r` without invoking the `fakeroot` command.   So that approach isn't going to work.  Instead, when processing the `--fakeroot` option I think it will need to detect if `unshare -r` (or its equivalent) had already been done and if so skip calling `fakeroot.UnshareRootMapped` but still invoke the `fakeroot` command.

JasonYangShadow commented:
> I think I know what happened. The removal of the `--fakeroot` flag resulted in it only doing the `unshare -r` equivalent without invoking the `fakeroot` command. So that approach isn't going to work. Instead, when processing the `--fakeroot` option I think it will need to detect if `unshare -r` (or its equivalent) had already been done and if so skip calling `fakeroot.UnshareRootMapped` but still invoke the `fakeroot` command.

is this line doing what you mentioned above?
https://github.com/JasonYangShadow/apptainer/blob/issue/2189/internal/pkg/runtime/launch/launcher_linux.go#L103

DrDaveD commented:
Yes it looks like that is already being done.  So we're back to not understanding how to fix the problem, aren't we ...

JasonYangShadow commented:
> Yes it looks like that is already being done. So we're back to not understanding how to fix the problem, aren't we ...

I updated the code and test case. Also added one test case.

DrDaveD commented on internal/pkg/runtime/launch/launcher_linux.go:
```diff
@@ -1073,9 +1076,10 @@ func (l *Launcher) setCgroups(instanceName string) error {
 	// If we are an instance, always use a cgroup if possible, to enable stats.
 	// root can always create a cgroup.
-	useCG := l.uid == 0
+	sylog.Debugf("During setting cgropus configuration, uid: %d, namespace.user: %t, fakeroot: %t, unprivileged: %t", l.uid, l.cfg.Namespaces.User, l.cfg.Fakeroot, namespaces.IsUnprivileged())
```
Typo `cgropus` -> `cgroups`

JasonYangShadow replied:
thanks, updated!

DrDaveD commented on internal/pkg/runtime/launch/launcher_linux.go:
```diff
@@ -118,7 +118,7 @@ func (l *Launcher) Exec(ctx context.Context, image string, args []string, instan
 		} else if (l.uid != 0) && (!fakeroot.IsUIDMapped(l.uid) || l.cfg.IgnoreSubuid) {
 			sylog.Infof("User not listed in %v, trying root-mapped namespace", fakeroot.SubUIDFile)
-			l.cfg.Fakeroot = false
+			l.cfg.Fakeroot = true
```
This is redundant because it's already inside of an `if` statement where the flag is true.  Just delete the line, although I'm not sure what difference it makes if it is false or true in this case.

JasonYangShadow replied:
thanks, updated!

DrDaveD commented on internal/pkg/runtime/launch/launcher_linux.go:
```diff
@@ -1084,6 +1088,7 @@ func (l *Launcher) setCgroups(instanceName string) error {
 		useCG = true
 	}
 
+	sylog.Debugf("During setting cgroups configuration, whether use cgroup manager: %t", useCG)
```
Rather than always including this statement, put a debug statement inside of `useCG` of "Using cgroups manager".  The absence of the message means it will be false.

JasonYangShadow replied:
thanks, updated!

JasonYangShadow commented on e2e/run/run.go:
```diff
@@ -402,15 +402,14 @@ func (c ctx) testAddPackageWithFakerootAndTmpfs(t *testing.T) {
 		e2e.WithCommand("exec"),
 		e2e.WithArgs("--no-home", "--writable", "--ignore-userns", "--ignore-subuid", sandbox, "sh", "-c", "apt-get update && apt-get install -y openssh-client"),
-		e2e.ExpectExit(100), // fails because only fakeroot is used. error with addgroup.
+		e2e.ExpectExit(0),
```
thanks! restored the `l.cfg.Fakeroot =false` line and that regression test case.

DrDaveD commented:
Please cherry-pick to release-1.3
